### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Angener**

### CODE REVIEW

File: module2/persistence/src/test/java/com/epam/esm/dao/InMemoryDbConfig.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+import java.sql.SQLException;
+
+@SpringJUnitWebConfig(JdbcConfig.class)
+@ContextConfiguration(
+        classes = {JdbcConfig.class},
+        loader = AnnotationConfigContextLoader.class)
+public class InMemoryDbConfig {
+    private static final String CREATE_TAG_TABLE = "CREATE TABLE IF NOT EXISTS tag (" +
```
Comment (aolshevskiy) [inline module2/persistence/src/test/java/com/epam/esm/dao/InMemoryDbConfig.java:18]: Why do we have this create table statements in code while we have init.sql in the connection url?

File: module2/service/src/main/java/com/epam/esm/config/ServiceConfig.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import com.epam.esm.service.GiftCertificateServiceImpl;
+import com.epam.esm.service.TagServiceImpl;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+@Configuration
+@Import(JdbcConfig.class)
+@ComponentScan(basePackageClasses = {TagServiceImpl.class, GiftCertificateServiceImpl.class})
```
Comment (aolshevskiy) [inline module2/service/src/main/java/com/epam/esm/config/ServiceConfig.java:12]: Root service package name would be shorter

File: module2/service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.experimental.FieldDefaults;
+
+import java.util.Set;
+
+@Data
+@NoArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class GiftCertificateDto {
+    GiftCertificate giftCertificate;
+    String[] fields;
```
Comment (aolshevskiy) [inline module2/service/src/main/java/com/epam/esm/dto/GiftCertificateDto.java:17]: I don't get the logic behind this array

File: module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.stream.Collectors;
+
+@Service
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    GiftCertificateDao giftCertificateDao;
+    TagService tagService;
+
+    Comparator<GiftCertificateWithTagsDto> dateComparator = Comparator.comparing(d ->
```
Comment (aolshevskiy) [inline module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java:32]: Please use SQL `ORDER BY` for sorting next time

File: module2/service/src/test/java/com/epam/esm/service/GiftCertificateServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@ExtendWith(SpringExtension.class)
+@SpringJUnitConfig({ServiceConfig.class})
+public class GiftCertificateServiceImplTest {
```
Comment (aolshevskiy) [inline module2/service/src/test/java/com/epam/esm/service/GiftCertificateServiceImplTest.java:39]: Still no unit test for the service

File: module2/web/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.support.ResourceBundleMessageSource;
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+
+@Configuration
+@Import(ServiceConfig.class)
+@EnableWebMvc
+@ComponentScan(basePackages = {"com.epam.esm.controller", "org.baeldung.web"})
```
Comment (aolshevskiy) [inline module2/web/src/main/java/com/epam/esm/config/SpringConfig.java:15]: `org.baeldung.web`, huh?

File: module2/web/src/main/java/com/epam/esm/controller/GIftCertificateController.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.io.IOException;
+
+@RestController
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class GIftCertificateController {
```
Comment (aolshevskiy) [inline module2/web/src/main/java/com/epam/esm/controller/GIftCertificateController.java:33]: `GIft` should be `Gift`

File: module2/web/src/main/java/com/epam/esm/exception/ExceptionDetail.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.experimental.FieldDefaults;
+import org.springframework.http.HttpStatus;
+
+
+@Getter
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+public enum ExceptionDetail {
```
Comment (aolshevskiy) [inline module2/web/src/main/java/com/epam/esm/exception/ExceptionDetail.java:13]: Enum is redundant. See above comment on SqlScript for details

**Review by aolshevskiy — APPROVED** [review]
Still no fast unit tests, but still acceptable.

File: module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+    private void checkUpdatableFields(String[] fields) throws UpdatingForbiddenFieldsException {
+        if (isUpdatableRequestContainsForbiddenFields(fields)) {
+            throw new UpdatingForbiddenFieldsException();
+        }
+    }
+
+    private boolean isUpdatableRequestContainsForbiddenFields(String[] fields) {
+        return Arrays.stream(fields)
+                .anyMatch(field -> field.trim().toLowerCase().equals("createdate") ||
```
Comment (alex9m) [inline module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java:59]: 'toLowerCase' +' equals' can be replaced by 'equalsIgnoreCase'

File: module2/gradlew [hunk 10]
```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env sh
```
Comment (dzmitrySazonov) [inline module2/gradlew:1]: do we need this file here?

File: module2/gradlew.bat [hunk 11]
```diff
@@ -0,0 +1,100 @@
+@rem
+@rem Copyright 2015 the original author or authors.
```
Comment (dzmitrySazonov) [inline module2/gradlew.bat:2]: do we need this file here?

File: module2/persistence/src/main/java/com/epam/esm/config/JdbcConfig.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+import javax.sql.DataSource;
+
+@Configuration
+@PropertySource("classpath:${spring.profiles.active}.properties")
+@ComponentScan(basePackages = {"com.epam.esm.dao"})
+@EnableTransactionManagement
+public class JdbcConfig {
+
+    @Value("${postgres.driverClass}") private String driverClass;
```
Comment (dzmitrySazonov) [inline module2/persistence/src/main/java/com/epam/esm/config/JdbcConfig.java:24]: code style

File: module2/persistence/src/main/java/com/epam/esm/config/JdbcConfig.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        return new JdbcTemplate(dataSource);
+    }
+
+    @Bean
+    public NamedParameterJdbcTemplate getNamedParameterJdbcTemplate(JdbcTemplate jdbcTemplate) {
+        return new NamedParameterJdbcTemplate((jdbcTemplate));
+    }
+
+    @Bean
+    public SimpleJdbcInsert getTag_GiftCertificateJdbcTemplate(JdbcTemplate jdbcTemplate) {
```
Comment (dzmitrySazonov) [inline module2/persistence/src/main/java/com/epam/esm/config/JdbcConfig.java:42]: naming

File: module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        return parameters;
+    }
+
+    private String prepareParameterForInsertingToSqlScript(String partNameOrDescription) {
+        return "%" + partNameOrDescription + "%";
+    }
+
+    @Override
+    @Transactional
+    public void update(GiftCertificate certificate, String[] fields, @Nullable Set<Tag> tags) {
```
Comment (dzmitrySazonov) [inline module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java:166]: better to return updated entity

File: module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+                                  JdbcTemplate jdbcTemplate,
+                                  TagDao tagDao) {
+        super(jdbcTemplate, namedParameterJdbcTemplate);
+        this.tagDao = tagDao;
+        this.simpleJdbcInsert = simpleJdbcInsert;
+    }
+
+    @Override
+    @Transactional(propagation = Propagation.REQUIRED)
+    public long save(GiftCertificate certificate, Set<Tag> tags) {
```
Comment (dzmitrySazonov) [inline module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDaoImpl.java:90]: better to return entity

File: module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDao.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.lang.Nullable;
+
+import java.util.List;
+import java.util.Set;
+
+public interface GiftCertificateDao {
+
+    long save(GiftCertificate certificate, Set<Tag> tags);
+
+    List<GiftCertificate> getAll();
```
Comment (dzmitrySazonov) [inline module2/persistence/src/main/java/com/epam/esm/dao/GiftCertificateDao.java:14]: the same methods as in TagDao, maybe it is better to move equal methods to common abstract Dao?

File: module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        return tagName.trim().length() > 0 || partOfName.trim().length() > 0;
+    }
+
+    private List<GiftCertificate> collect(SortCertificatesType type, String tagName, String partOfName) {
+        return (tagName.trim().length() > 0) ?
+                collectCertificatesByTagNameWithOrWithoutByPartNameOrDescription(type, tagName, partOfName)
+                : (collectByPartNameOrDescriptionOnly(type, partOfName));
+    }
+
+    private List<GiftCertificate> collectCertificatesByTagNameWithOrWithoutByPartNameOrDescription
```
Comment (dzmitrySazonov) [inline module2/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java:113]: TLDR

File: module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class GiftCertificateController {
+
+    GiftCertificateService service;
+    ObjectMapper mapper;
+
+    @PostMapping("/giftCertificates")
+    @ResponseStatus(HttpStatus.CREATED)
+    public long save(@RequestBody GiftCertificateDto dto) {
```
Comment (dzmitrySazonov) [inline module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:40]: return created entity in response

File: module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+            return service.save(dto);
+        } catch (DuplicateKeyException ex) {
+            throw new LocalizedControllerException(ExceptionDetail.NAME_IS_NOT_UNIQUE);
+        } catch (DataIntegrityViolationException ex) {
+            throw new LocalizedControllerException(ExceptionDetail.CERTIFICATE_TAGS_IS_NOT_AVAILABLE);
+        }
+    }
+
+    @PatchMapping("/giftCertificates")
+    public void update(@RequestBody GiftCertificateDto dto) {
```
Comment (dzmitrySazonov) [inline module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:51]: return updated entity in response

File: module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            service.update(dto);
+        } catch (UpdatingForbiddenFieldsException ex) {
+            throw new LocalizedControllerException(ExceptionDetail.UPDATING_FORBIDDEN_DATE_FIELDS);
+        } catch (DuplicateKeyException ex) {
+            throw new LocalizedControllerException(ExceptionDetail.NAME_IS_NOT_UNIQUE);
+        }
+    }
+
+    @GetMapping("/giftCertificates")
+    public String getAll() throws IOException {
```
Comment (dzmitrySazonov) [inline module2/web/src/main/java/com/epam/esm/controller/GiftCertificateController.java:62]: why string? return object

File: module2/web/src/main/java/com/epam/esm/controller/TagController.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+        try {
+            return service.getById(id);
+        } catch (EmptyResultDataAccessException ex) {
+            throw new LocalizedControllerException(ExceptionDetail.TAG_NOT_FOUND);
+        }
+    }
+
+    @PostMapping("/tags")
+    @ResponseStatus(HttpStatus.CREATED)
+    public long save(@RequestBody Tag tag) {
```
Comment (dzmitrySazonov) [inline module2/web/src/main/java/com/epam/esm/controller/TagController.java:50]: return entity

File: module2/gradlew (hunk 10 above)
Comment (Angener) [reply module2/gradlew:1]: I believe the file is needed into the project structure. Also, I had uploaded the .gitignore file from the official repository.

File: module2/gradlew.bat (hunk 11 above)
Comment (Angener) [reply module2/gradlew.bat:2]: It seems we need

**Pull request merged**